### PR TITLE
Fixes saml login flow to work with anonymous auth

### PR DESCRIPTION
**Pull request by DarshitChanpura**
### Description
This PR fixes 2 things:
1. Anonymous login button would not show if only 1 auth option was supplied
2. Broken SAML auth flow when anonymous auth is enabled.
3. Automatic login as anonymous user upon SAML user logout. (Discovered while fixing this bug)

### Category
- Bug fix

### Why these changes are required?
- For SAML and anonymous auth options to function when both options are enabled
- To prevent auto login as anonymous user upon SAML user logout

### What is the old behavior before changes and new behavior after changes?
- At present, SAML auth flow is broken since anonymous auth automatically assumes the Anonymous user when no auth credentials are found. See security plugin PR[1] for more details.

[1] - Companion PR: https://github.com/opensearch-project/security/pull/4152

### Issues Resolved
- Resolves https://github.com/opensearch-project/security-dashboards-plugin/issues/1731
- Resolves https://github.com/opensearch-project/security-dashboards-plugin/issues/1840

### Testing
- Automated + Manual

Manual testing:

https://github.com/opensearch-project/security-dashboards-plugin/assets/35282393/e40e974d-1141-47a7-9b2e-2006b5ca0981

### Check List
- [x] New functionality includes testing
~- [ ] New functionality has been documented~
- [x] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: server/auth/types/basic/basic_auth.ts
Comment (DarshitChanpura) [inline server/auth/types/basic/basic_auth.ts:128]: This if block caused all logout requests to be redirected to the anonymous login page when anonymous auth is enabled. This is incorrect. All logout requests should redirect to login page. (Similar change below for multi_auth.ts)

File: test/constant.ts
Comment (DarshitChanpura) [inline test/constant.ts:21]: unused. Discovered while local run of basic auth tests.

File: server/auth/types/openid/routes.ts
Comment (derek-ho) [inline server/auth/types/openid/routes.ts:191]: Do we need to do this for OpenID? If we do, do we need to add it to the refresh token flow? https://github.com/opensearch-project/security-dashboards-plugin/blob/main/server/auth/types/openid/openid_auth.ts#L303

Comment (derek-ho) [reply server/auth/types/openid/routes.ts:191]: Do we want to add it for all the other auth types as well, such as proxy auth?

Comment (DarshitChanpura) [reply server/auth/types/openid/routes.ts:191]: It is a place holder, and adding it here should suffice. OpenId directly calls the openid url and doesn't rely on error response from backend, like SAML, to redirect to the IdP. Hence, instead of `undefined` I chose to add a placeholder.

**Comment by DarshitChanpura** [comment]
should be merged post https://github.com/opensearch-project/security/pull/4152 is merged.

File: public/apps/login/test/login-page.test.tsx
Comment (derek-ho) [inline public/apps/login/test/login-page.test.tsx:226]: is there a constant you can use for openid?

File: server/auth/types/basic/basic_auth.ts
Comment (derek-ho) [reply server/auth/types/basic/basic_auth.ts:128]: Nice!

File: server/backend/opensearch_security_client.ts
Comment (derek-ho) [inline server/backend/opensearch_security_client.ts:229]: Nit: I saw this on other APIs as well, can we avoid to have multiple possible undefined/any parameters and instead replace this with a typed/named object? It might get confusing with how to call something where several of the later options are optional or undefined. Not sure if that should be addressed here or we need another issue for repo cleanup.

**Review by derek-ho — COMMENTED** [review]
Left a few comments, but mostly LGTM. Will approve once we get alignment on the backend change as well, since this relies on that.

**Comment by derek-ho** [comment]
@DarshitChanpura is this ready for review? Are you expecting integration tests fail until the backend PR is merged?

**Comment by DarshitChanpura** [comment]
`jwt_auth.test.ts` integration test is failing when run on main as well. Not sure what changed between last commit on main and this commit.

**Comment by DarshitChanpura** [comment]
<details>
<summary>Cypress SAML E2E: </summary>

```
➜  security-dashboards-plugin git:(fixes-saml-login-flow) ✗ yarn cypress:run --headed --env loginMethod=saml_multiauth --spec ./test/cypress/e2e/saml/saml_auth_test.spec.js
yarn run v1.22.19
$ cypress run --headed --env loginMethod=saml_multiauth --spec ./test/cypress/e2e/saml/saml_auth_test.spec.js

DevTools listening on ws://127.0.0.1:64052/devtools/browser/7ac4e92d-1f4f-415b-a0c9-6a4696838692
2024-04-08 23:37:04.866 Cypress[16305:1323709] WARNING: Secure coding is not enabled for restorable state! Enable secure coding by implementing NSApplicationDelegate.applicationSupportsSecureRestorableState: and returning YES.

====================================================================================================

  (Run Starting)

  ┌────────────────────────────────────────────────────────────────────────────────────────────────┐
  │ Cypress:        13.6.0                                                                         │
  │ Browser:        Electron 114                                                                   │
  │ Node Version:   v18.19.0 (./.nvm/versions/node/v18.19.0/bin/node)                  │
  │ Specs:          1 found (saml_auth_test.spec.js)                                               │
  │ Searched:       test/cypress/e2e/saml/saml_auth_test.spec.js                                   │
  └────────────────────────────────────────────────────────────────────────────────────────────────┘


────────────────────────────────────────────────────────────────────────────────────────────────────
                                                                                                    
  Running:  saml_auth_test.spec.js                                                          (1 of 1)
Browserslist: caniuse-lite is outdated. Please run:
  npx update-browserslist-db@latest
  Why you should do it regularly: https://github.com/browserslist/update-db#readme


  Log in via SAML
    ✓ Login to app/opensearch_dashboards_overview#/ when SAML is enabled (17254ms)
    ✓ Login to app/dev_tools#/console when SAML is enabled (2353ms)
    ✓ Login to Dashboard with Hash (2070ms)
    ✓ Tenancy persisted after logout in SAML (10986ms)
    ✓ Login to Dashboard with Goto URL (3076ms)


  5 passing (36s)


  (Results)

  ┌────────────────────────────────────────────────────────────────────────────────────────────────┐
  │ Tests:        5                                                                                │
  │ Passing:      5                                                                                │
  │ Failing:      0                                                                                │
  │ Pending:      0                                                                                │
  │ Skipped:      0                                                                                │
  │ Screenshots:  0                                                                                │
  │ Video:        false                                                                            │
  │ Duration:     36 seconds                                                                       │
  │ Spec Ran:     saml_auth_test.spec.js                                                           │
  └────────────────────────────────────────────────────────────────────────────────────────────────┘


====================================================================================================

  (Run Finished)


       Spec                                              Tests  Passing  Failing  Pending  Skipped  
  ┌────────────────────────────────────────────────────────────────────────────────────────────────┐
  │ ✔  saml_auth_test.spec.js                   00:36        5        5        -        -        - │
  └────────────────────────────────────────────────────────────────────────────────────────────────┘
    ✔  All specs passed!                        00:36        5        5        -        -        -  

✨  Done in 40.49s.
➜  security-dashboards-plugin git:(fixes-saml-login-flow) ✗ 
```
</details>

**Comment by DarshitChanpura** [comment]
all failures like this: `"message": "[illegal_argument_exception] request [/_plugins/_security/authinfo] contains unrecognized parameter: [auth_type]"` will be solved post merge of https://github.com/opensearch-project/security/pull/4152

**Comment by DarshitChanpura** [comment]
Cypress failures seem related to this: https://github.com/cypress-io/github-action/issues/1136.

They seem to get stuck at:
```
Warning: vkCreateInstance: Found no drivers!
Warning: vkCreateInstance failed with VK_ERROR_INCOMPATIBLE_DRIVER
    at CheckVkSuccessImpl (../../third_party/dawn/src/dawn/native/vulkan/VulkanError.cpp:88)
    at CreateVkInstance (../../third_party/dawn/src/dawn/native/vulkan/BackendVk.cpp:458)
    at Initialize (../../third_party/dawn/src/dawn/native/vulkan/BackendVk.cpp:344)
    at Create (../../third_party/dawn/src/dawn/native/vulkan/BackendVk.cpp:266)
    at operator() (../../third_party/dawn/src/dawn/native/vulkan/BackendVk.cpp:521)
```

File: server/auth/types/authentication_type.ts
Comment (scrawfor99) [inline server/auth/types/authentication_type.ts:122]: Do we need to make the addition of this the default behavior?

File: server/auth/types/authentication_type.ts
Comment (scrawfor99) [inline server/auth/types/authentication_type.ts:213]: Should probably standardize the authType string use to be a enum with constants.

**Review by scrawfor99 — DISMISSED** [review]
Overall looks good, two small comments.

File: server/auth/types/authentication_type.ts
Comment (DarshitChanpura) [reply server/auth/types/authentication_type.ts:122]: yes, it is required as an argument for `authinfo` method. '' here since this block will only be called on rest-API calls only, and for rest calls made as Anonymous user, the backend will evaluate this as `opendistro_security_anonymous` since no auth header would have been passed, so '' will work.

File: server/auth/types/authentication_type.ts
Comment (DarshitChanpura) [reply server/auth/types/authentication_type.ts:213]: good catch. I've pushed a new commit by replacing this magic constant.

File: server/backend/opensearch_security_client.ts
Comment (derek-ho) [inline server/backend/opensearch_security_client.ts:75]: Just curious - is there a reason you put this in a list? Or is that syntax for something else? Why can't it just be AUTH_TYPE_PARAM?

File: server/backend/opensearch_security_plugin.ts
Comment (derek-ho) [inline server/backend/opensearch_security_plugin.ts:50]: Is this part necessary? There isn't validation on any other routes? Additionally there is some code here that we may want to test or use a helper library like lodash to do this?

File: server/backend/opensearch_security_client.ts
Comment (DarshitChanpura) [reply server/backend/opensearch_security_client.ts:75]: this is not a list, it evaluate the constant: `AUTH_TYPE_PARAM ` to be `auth_type`. Without this it would literally set `AUTH_TYPE_PARAM` as the key.

File: server/backend/opensearch_security_plugin.ts
Comment (DarshitChanpura) [reply server/backend/opensearch_security_plugin.ts:50]: yes, without this the url is not modified to pass `?auth_type=` query param

File: server/readonly/readonly_service.ts
Comment (derek-ho) [inline server/readonly/readonly_service.ts:112]: Shouldn't cookie always exist at this point? Can we clean this up instead of chaining ?'s? I got slightly confused reading this

Comment (DarshitChanpura) [reply server/readonly/readonly_service.ts:112]: if you look at the code above, cookie is fetched via `const cookie = await this.securitySessionStorageFactory.asScoped(request).get();` which has a chance that it will return null if session storage was cleared. I think we can leave it as is.

File: server/backend/opensearch_security_plugin.ts
Comment (derek-ho) [reply server/backend/opensearch_security_plugin.ts:50]: Ok if its necessary can you extract this out to a helper function and write a test for it?

Comment (DarshitChanpura) [reply server/backend/opensearch_security_plugin.ts]: sure, I have addressed this in the latest commit.

**Review by derek-ho — DISMISSED** [review]
You are the 🐐

File: server/backend/opensearch_security_client.ts
Comment (DarshitChanpura) [reply server/backend/opensearch_security_client.ts:197]: only this will work, as all the following requests will have Authorization header (from cookie), and if they don't the request will be redirected to login page. This is what my understanding is.

File: server/auth/types/saml/routes.ts
Comment (DarshitChanpura) [reply server/auth/types/saml/routes.ts:138]: I have used to set it explicitly to avoid redirecting auto-login as anonymous upon login failure. This was the issue I faced with basic auth when passing wrong credentials led to auto login as anonymous because backend couldn't differentiate anonymous from non-anonymous logon requests since front-end didn't specify it in particular.

File: server/auth/types/basic/basic_auth.ts
Comment (cwperks) [inline server/auth/types/basic/basic_auth.ts:115]: When will `ANONYMOUS_AUTH_LOGIN` be called if this line is removed?

Comment (cwperks) [reply server/auth/types/basic/basic_auth.ts:115]: Oh nvm, I see this is contained in `handleUnauthedRequest`. Does that mean its expected to log in as anonymous after getting an authentication failure?

File: server/auth/types/saml/routes.ts
Comment (cwperks) [reply server/auth/types/saml/routes.ts:138]: I think its possible to pass only where necessary: https://github.com/cwperks/security-dashboards-plugin/pull/6

There's a separate issue on the log out flow for multi-auth where the logic for automatically redirect to anonymous for multi-auth is unnecessary: https://github.com/opensearch-project/security-dashboards-plugin/blob/main/server/auth/types/multiple/multi_auth.ts#L169-L176

I think its unnecessary because when multi-auth is enabled, login with anonymous should always be initiated by pushing the button `Login as Anonymous`. From what I understand, this block is necessary in the [basic_auth handler](https://github.com/opensearch-project/security-dashboards-plugin/blob/main/server/auth/types/basic/basic_auth.ts#L114-L121) because security-dashboards-plugin needs to support the case where anonymous login is the _only_ form of login.

Comment (DarshitChanpura) [reply server/auth/types/saml/routes.ts:138]: I tried this, and without passing it explicitly here, the login threw 500 error when i logged out and tried to log in again.

File: server/auth/types/basic/basic_auth.ts
Comment (DarshitChanpura) [reply server/auth/types/basic/basic_auth.ts]: I;ve reverted this change to keep it simple.

File: server/auth/types/saml/routes.ts
Comment (DarshitChanpura) [reply server/auth/types/saml/routes.ts:138]: I've removed all changes out of scope of this bug fix to keep the review much simpler.

File: server/backend/opensearch_security_client.ts
Comment (derek-ho) [reply server/backend/opensearch_security_client.ts:187]: Maybe I didn't get this change completely - shouldn't the backend only be checking for "anonymous"? Meaning this wouldn't have any effect, even without the authorization header?

Comment (derek-ho) [reply server/backend/opensearch_security_client.ts:187]: Specifically this block:

```
if (params.containsKey("auth_type")) {
            return params.get("auth_type").equals("anonymous");
        }
        return !headers.containsKey(HttpHeaders.AUTHORIZATION);
```

should return false even if SAML doesn't pass it in, right?

Comment (DarshitChanpura) [reply server/backend/opensearch_security_client.ts:187]: `[AUTH_TYPE_PARAM]: AuthType.SAML` is required to differentiate the normal unauth request (anonymous) from SAML unauth request, since SAML expects backend to pass the redirect route. without this differentiation, SAML request would be considered as anonymous accidentally by backend while would lead to 500 being thrown.

Comment (derek-ho) [reply server/backend/opensearch_security_client.ts:187]: Code like this on the backend:

```
private boolean isRequestForAnonymousLogin(Map<String, String> params, Map<String, List<String>> headers) {
 helper = !headers.containsKey(HttpHeaders.AUTHORIZATION);
        if (params.containsKey("auth_type")) {
            return helper && params.get("auth_type").equals("anonymous");
        }
        return helper;
    }
    ```

Or its logical equivalent would make passing in SAML unecessary

**Review by derek-ho — APPROVED** [review]

**Review by cwperks — APPROVED** [review]
LGTM! Thank you @DarshitChanpura!

FYI I don't think there is a way to setup Basic Auth + Anonymous without it automatically logging in as anonymous when first loading OSD. I think the issue is in [auth_handler_factory.ts](https://github.com/cwperks/security-dashboards-plugin/commit/dd12e76d7d0bedf3a0538ddab37bb14250a948fc) where even if multi-auth is enabled and only one auth_type is in the list it will create an instance of the single auth type instead of using [MultipleAuthentication](https://github.com/opensearch-project/security-dashboards-plugin/blob/main/server/auth/types/multiple/multi_auth.ts#L34)

**Pull request merged**
